### PR TITLE
Port grouping redo/v13

### PR DESCRIPTION
**Pull request by inashivb**
Fixes and cleanups to https://github.com/OISF/suricata/pull/10466

Still TODO:
- cppcheck issues
- optimization of whitelist grouping

SV_BRANCH=https://github.com/OISF/suricata-verify/pull/1658

Note: I think https://redmine.openinfosecfoundation.org/issues/6414 does not quite say what this PR does even though the PR takes care of it. I think we need a separate ticket for the optimization changes done here. Thoughts?

### CODE REVIEW

**Comment by victorjulien** [comment]
suggesting
```diff
diff --git a/src/detect-engine-build.c b/src/detect-engine-build.c
index 53b99f48cb..d112698719 100644
--- a/src/detect-engine-build.c
+++ b/src/detect-engine-build.c
@@ -1282,7 +1282,7 @@ static DetectPort *RulesGroupByPorts(DetectEngineCtx *de_ctx, uint8_t ipproto, u
 
     uint8_t *unique_port_points = (uint8_t *)SCCalloc(MAX_PORT_VAL, sizeof(uint8_t));
     if (unique_port_points == NULL)
-        goto error;
+        return NULL;
 
     while (s) {
         /* IP Only rules are handled separately */
diff --git a/src/util-interval-tree.c b/src/util-interval-tree.c
index cef8d47773..911010843b 100644
--- a/src/util-interval-tree.c
+++ b/src/util-interval-tree.c
@@ -165,7 +165,6 @@ static void FindOverlaps(DetectEngineCtx *de_ctx, const uint16_t port, const uin
         SCIntervalNode *root, DetectPort **list)
 {
     DetectPort *new_port = NULL;
-    int cnt_overlaps = 0;
     int stack_depth = 0;
     SCIntervalNode *stack[100], *current = root;
     memset(&stack, 0, sizeof(stack));
@@ -177,10 +176,6 @@ static void FindOverlaps(DetectEngineCtx *de_ctx, const uint16_t port, const uin
                 break;
             }
             const bool is_overlap = IsOverlap(port, port2, current);
-            if (is_overlap) {
-                cnt_overlaps++;
-            }
-
             if (is_overlap && (new_port == NULL)) {
                 /* Allocate memory for port obj only if it's first overlap */
                 new_port = DetectPortInit();
```

**Comment by suricata-qa** [comment]
Information:
 

ERROR: QA failed on SURI_TLPW2_autofp_suri_time.
 | field | baseline | test | % |
|--|--|--|--| 
| | SURI_TLPW2_autofp_stats_chk | | |
| .uptime | 101 | 112 | 110.89% | 

Pipeline 18645

**Comment by suricata-qa** [comment]
Information:
 

ERROR: QA failed on SURI_TLPW2_autofp_suri_time.
 | field | baseline | test | % |
|--|--|--|--| 
| | SURI_TLPW2_autofp_stats_chk | | |
| .uptime | 101 | 118 | 116.83% |
| | SURI_TLPR1_stats_chk | | |
| .uptime | 651 | 678 | 104.15% | 

Pipeline 18646

File: src/util-interval-tree.h [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+#define __UTIL_INTERVAL_TREE_H__
+
+#include "interval-tree.h"
+#include "detect.h"
+
+typedef struct SCIntervalNode {
+    uint16_t port;  /* low port of a port range */
+    uint16_t port2; /* high port of a port range */
+    uint16_t max;   /* max value of the high port in the subtree rooted at this node */
+    uint8_t flags;  /* port specific flags */
```
Comment (inashivb) [inline src/util-interval-tree.h:34]: Not really used anywhere until the list is created so remove

File: src/util-interval-tree.c [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+                    DEBUG_VALIDATE_BUG_ON(new_port->port < (*list)->last->port);
+                    (*list)->last->next = new_port;
+                    (*list)->last = new_port;
+                } else {
+                    SCLogDebug("Port already exists in the list");
+                    goto error;
+                }
+            } else if (is_overlap && (new_port != NULL)) {
+                SCLogDebug("find overlaps for [%u:%u]: adding sigs to new_port [%u:%u]", port,
+                        port2, port, port2);
```
Comment (inashivb) [inline src/util-interval-tree.c:209]: same ports printed in debug stmt

File: src/util-interval-tree.c [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+ *
+ * \param port Given low port
+ * \param port2 Given high port
+ * \param ptr Pointer to the node in the tree to be checked against
+ *
+ * \return true if an overlaps was found, false otherwise
+ */
+static bool IsOverlap(const uint16_t port, const uint16_t port2, const SCIntervalNode *ptr)
+{
+    if (port <= ptr->port2 && ptr->port <= port2) {
```
Comment (inashivb) [inline src/util-interval-tree.c:141]: explain why

**Pull request closed without merging**